x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -226,12 +231,36 @@ public void addCompletionListener(Consumer<AsyncSearchResponse> listener) {
             }
         }
         if (executeImmediately) {
-            var response = getResponseWithHeaders();
-            try {
-                listener.accept(response);
-            } finally {
-                response.decRef();
-            }
+            getResponseWithHeaders(new ActionListener<>() {
```

Comment:
this would apply for all action listeners we create - should we use `ActionListener.wrap` to make sure that if `onResponse` throws an Exception `onFailure` gets called?